### PR TITLE
*: update kvrpc response

### PR DESCRIPTION
**Pull request by disksing**
@siddontang @ngaut @morefreeze

### CODE REVIEW

File: src/server/kv.rs [hunk 1]
```diff
@@ -416,50 +345,94 @@ impl StoreHandler {
...
-        res_type.set_leader_info(err.get_not_leader().to_owned());
-    } else {
-        error!("{:?}", err);
-        res_type.set_field_type(ResultType_Type::Other);
-        res_type.set_msg(format!("engine error: {:?}", err));
-    }
-    item.set_res_type(res_type);
-    item
+fn extract_region_error<T>(res: &StorageResult<T>) -> Option<RegionError> {
+    use storage::Error as StorageError;
```
Comment (ngaut) [inline src/server/kv.rs]: Redundant code. Should be clean up.

**Comment by disksing** [comment]
PTAL @siddontang @ngaut @morefreeze

**Comment by ngaut** [comment]
PTAL @BusyJay

File: src/server/kv.rs [hunk 2]
```diff
@@ -200,196 +203,125 @@ impl StoreHandler {
...
-        resp.set_cmd_get_resp(cmd_get_resp);
+        resp.set_cmd_get_resp(get_resp);
         resp
     }
 
     fn cmd_scan_done(kvs: StorageResult<Vec<StorageResult<KvPair>>>) -> Response {
         let mut resp = Response::new();
-        let mut cmd_scan_resp = CmdScanResponse::new();
-        cmd_scan_resp.set_ok(kvs.is_ok());
         resp.set_field_type(MessageType::CmdScan);
```
Comment (BusyJay) [inline src/server/kv.rs]: Do we have to set the field type?

File: src/server/kv.rs [hunk 3]
```diff
@@ -200,196 +203,125 @@ impl StoreHandler {
 
     fn cmd_get_done(r: StorageResult<Option<Value>>) -> Response {
         let mut resp = Response::new();
-        let mut cmd_get_resp = CmdGetResponse::new();
-        let mut res_type = ResultType::new();
+        resp.set_field_type(MessageType::CmdGet);
+        if let Some(e) = extract_region_error(&r) {
+            resp.set_region_error(e);
+            return resp;
+        }
```
Comment (BusyJay) [inline src/server/kv.rs]: I think L205~L210 can be moved to `self.make_cb`.

**Comment by BusyJay** [comment]
rest LGTM

File: src/server/kv.rs (hunk 2 above)
Comment (disksing) [inline src/server/kv.rs]: Yes. Client will check it is matched.

**Comment by disksing** [comment]
@BusyJay PTAL

**Comment by ngaut** [comment]
Much cleaner.
PTAL @morefreeze

File: src/server/kv.rs [hunk 4]
```diff
@@ -527,20 +452,15 @@ mod tests {
...
-                   *actual_kvs[1].get_res_type());
-        assert_eq!(k1, actual_kvs[1].get_key());
-        assert_eq!(v1, actual_kvs[1].get_value());
+        let resp = build_resp(Ok(kvs), StoreHandler::cmd_scan_done);
+        assert_eq!(MessageType::CmdScan, resp.get_field_type());
+        let cmd = resp.get_cmd_scan_resp();
+        let pairs = cmd.get_pairs();
+        assert_eq!(2, pairs.len());
+        assert_eq!(k0, pairs[0].get_key());
+        assert_eq!(v0, pairs[0].get_value());
```
Comment (morefreeze) [inline src/server/kv.rs]: I think you should need check `pairs[0].get_error()`.

File: src/server/kv.rs [hunk 5]
```diff
@@ -416,108 +302,147 @@ impl StoreHandler {
...
+    let mut key_error = KeyError::new();
+    match *err {
+        StorageError::Txn(TxnError::Mvcc(MvccError::KeyIsLocked { ref key, ref primary, ts })) => {
+            let mut lock_info = LockInfo::new();
+            lock_info.set_key(key.to_owned());
+            lock_info.set_primary_lock(primary.to_owned());
+            lock_info.set_lock_version(ts);
+            key_error.set_locked(lock_info);
+        }
+        StorageError::Txn(TxnError::Mvcc(MvccError::WriteConflict)) => {
```
Comment (morefreeze) [inline src/server/kv.rs]: Other errors?

**Comment by BusyJay** [comment]
LGTM

**Comment by disksing** [comment]
PTAL @morefreeze

**Comment by morefreeze** [comment]
Please take my knees.
LGTM

**Pull request merged**